### ISSUE TITLE
Wrong entrypoint (hard coded?)

### ISSUE THREAD

**Issue by kaithar** [radare/radare2#12848]
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Multiple x86
| File format of the file you reverse (mandatory)      | See below
| Architecture/bits of the file (mandatory)            | See below
| r2 -v full output, **not truncated** (mandatory)     | Multiple, but we'll go with r2-3.1.3-140-g2462653a2

Courtesy of `i`:
```
format   smd
type     ROM
arch     m68k
bits     16
class    SEGA GENESIS    (C)T-48 1990.JUL
endian   little
machine  Sega Megadrive
os       smd
```
(I'm not sure why it insists it's a LE system, I have told it very clearly that it's BE and it's loading it as such, but that's not actually related to my bug)

### Expected behavior

In a properly formatted ROM, the entry point is specified by the dword at address 0x04.  In the case of the file I'm looking at, per that r_bin output, those 4 bytes contain `0x0001a000` ... that's pretty far in to the memory, but it is indeed the boilerplate entry point.  So the expected behaviour is that the analyser would read that and set entry0 to that address.

### Actual behavior

The entry point is being set to `0x00000200` instead.  I can kinda understand it, since what documentation I can find that mentions the entry point says it's "usually" at that address.  It makes sense but still seems wrong.  I'm guessing that somewhere in the analyser logic, that value is hard coded, but I don't know what component that's actually going to be in.

### Steps to reproduce the behavior 
I think it's probably safer if I don't share the ROM here, not really worth the risk.  Instead I'll share an abridged version.  
With the power of xxd, I've extracted the first 600 bytes of the file plus the first 16 bytes of the entry point.  The spaces in between are padded out with null bytes, so the entry is at the correct address.  Obviously it's not going to work in an emulator, but it should do for a test case?

[header_only.md.gz](https://github.com/radare/radare2/files/2776446/header_only.md.gz)

**Comment by Maijin** [comment]
For infor radare2 is an hexadecimal editor thus you don't need to use `xxd` you can open r2 in raw mode using `2 -n lol.exe` if you don't want r2 to map anything and use `px`.

Code for entry point detection is here: https://github.com/radare/radare2/blob/master/libr/bin/p/bin_smd.c#L293 feel free to improve it 👍

**Comment by kaithar** [comment]
Yeah, xxd was just handy for going "grab these bytes from this file and patch them in to the other" 

I'm not quite sure what values that function should be returning, it mentions 3 but I only see two in the return object and they're both set the same.
Hmmm, NMI, Reset and IRQ... which IRQ though?  They'll all probably be in SMD_Vectors which should be loaded from 0x0 with SMD_Header being at 0x100 ... no idea if the buf is already zero'd to that spot by the time that function gets called.

Honestly, the values in RomStart and co are garbage.   Those four are generally going to be `0x0`, rom size, `0x00ff0000`, `0x00ffffff` in the legit files... the rom size will be something like 0x03ffff, 0x07ffff, or 0x0fffff

**Comment by Maijin** [comment]
Feel free to submit a PR with the change you suggest.

**Comment by lowlyw** [comment]
i made this change: 

```diff
diff --git a/libr/bin/p/bin_smd.c b/libr/bin/p/bin_smd.c
index 6c9e31fac..f5613c45b 100644
--- a/libr/bin/p/bin_smd.c
+++ b/libr/bin/p/bin_smd.c
@@ -300,14 +300,16 @@ static RList *entries(RBinFile *bf) { // Should be 3 offsets pointed by NMI, RES
        if (!(ptr = R_NEW0 (RBinAddr))) {
                return ret;
        }
-#if 0
-       SMD_Header hdr = {{0}};
-       r_buf_read_at (bf->buf, 0, (ut8*)&hdr, sizeof (hdr));
-       ut64 baddr = 0; // r_read_be32 (&hdr.RomStart);
-       ptr->paddr = ptr->vaddr = baddr + 0x100 + sizeof (SMD_Header);
-#endif
-       ptr->paddr = ptr->vaddr = 0x100 + sizeof (SMD_Header);
-       r_list_append (ret, ptr);
+       if (bf->size < sizeof (SMD_Vectors)) {
+               eprintf ("ERR: binfile too small!\n");
+               ptr->paddr = ptr->vaddr = 0x100 + sizeof (SMD_Header);
+               r_list_append (ret, ptr);
+       } else {
+               SMD_Vectors vectors;
+               r_buf_read_at (bf->buf, 0, (ut8*)&vectors, sizeof (vectors));
+               ptr->paddr = ptr->vaddr = r_read_be32(&vectors.Reset);
+               r_list_append (ret, ptr);
+       }
        return ret;
 }
```

it makes this case correct.

but i have no idea if this is correct in all cases.

i downloaded sanic rom, but the entrypoint appears as this `vaddr=0x00000206`

```
            0x00000200      4e71           nop                         ; [02] -r-x section size 523776 n
            0x00000202      4e71           nop
            0x00000204      60fa           bra.b sym.IRQ1              ;[1]
            ;-- entry0:
            ;-- Reset:
            0x00000206      4ab900a10008   tst.l 0xa10008.l
        ┌─< 0x0000020c      6606           bne.b 0x214                 ;[2]
        │   0x0000020e      4a7900a1000c   tst.w 0xa1000c.l
```

is this right? if it is, then good, if not then wtf?

**Comment by kaithar** [comment]
Yup, that looks like the right entry point.  Thanks :)
